### ISSUE TITLE
Created C++ TMX Loader

### ISSUE THREAD

**Issue by martygrant** [bjorn/tiled#1128]
Hello, I created my own .tmx loader in C++ for use in some hobby projects. It uses RapidXML for file parsing, when exporting your tile maps in XML format. 

I hope I can have it added to your documentation site so it can be of some use to someone.

I have created a group of classes which form a data structure to keep a map in memory for use in a game/engine. The structure can be used directly, or used to build proprietary level objects more easily.

Here's a quick example of code usage:

```c++
TMXLoader* loader = new Loader();
loader->loadMap("testmap", "Assets/testmap.tmx"); 
loader->printMapData("testmap"); // prints map data to console, useful for debugging

loader->getMap("testmap"); // gets a map loaded and held by TMXLoader
loader->getMap("testmap")->getVersion();
loader->getMap("testmap")->getOrientation();
loader->getMap("testmap")->getWidth();
loader->getMap("testmap")->getHeight();
loader->getMap("testmap")->getTileWidth();
loader->getMap("testmap")->getTileHeight();
```

The TMXLoader class holds a container of TMXMaps, so a collection of maps can be loaded and stored, and be parsed at any time later on. 

What's supported:
* .tmx Maps exported as XML
* Multiple TileSets
* Multiple TileLayers
* User-defined properties for tilesets, layers and individual tiles

What's not supported yet:
* Object layers
* Terrain
* Tile flipping
* Image Layers

There is more information on my website here: http://www.midnightpacific.com/portfolio/tmxloader-for-tiled-map-editor/ where there is a link to my Bitbucket repository/link to clone from git.

**Comment by RySchmitt** [comment]
Your repository link is currently private, but it would be interesting to compare your API to the alternatives.

**Comment by martygrant** [comment]
Totally forgot to make it public. Thanks for pointing that out, I've changed the settings: https://bitbucket.org/martingrant/tmxloader

**Comment by bjorn** [comment]
Since there are already a large number of C++(11) map loaders listed, and some are actually pretty feature-complete, I have to ask why another one?

**Comment by martygrant** [comment]
I wanted to have an interface that works well for me, so created my own. It would be my first contribution to someone else's project so there is likely some problems and design issues in the code, but seems to work fine for my needs.

Mines differs slightly from the others. In [Tomáš Andrle's](http://blog.catnapgames.com/2011/10/10/simple-tmx-tilemap-parser/) the map must be read in the .CSV format, mines use .XML.

In loaders created by [halsafar's](https://github.com/halsafar/libtmx-parser), [andrewrk](https://github.com/andrewrk/tmxparser), [jube](https://github.com/jube/libtmx), and [solar-storm](https://github.com/solar-storm-studios/TMXParser) to access each part of a map you must create a reference, and use the reference. E.g you create a reference to a layer within a map, and use that reference to access layer data. My class gives you direct access to any part of a map, by passing the name. For example, to get the width of a layer: loader->getMap("map1")->getTileLayer("defaultLayer")->getWidth();

Also, andrewrk and jube's loaders there are more than 1 dependencies required (including Boost which I can find a pain to include in a project since its a huge library). Mines only requires RapidXML which is just a couple of header files, included in my repository.

Also, in the other loaders, they are designed to load a single map specifically. My class is able to load and them keep any and all maps you load, stored internally in the class so you can access them later at any time. The memory is cleaned up safely later. 

My loader is most similar to andrewrk, jube and solar-storm's. An additional difference with solar-storm's is any a user has public access to map data, meaning the data of a map once loaded can be altered. My class is read-only. There are also two classes required to fully load all the data associated with a map, for solar-storm's.

Hope if any of the people I mention read this I am not bashing their code in anyway! Just pointing out differences between the loader I wrote and theirs.

However, bjorn if you find there is no benefit in adding mines to the list then that's okay, I will continue using my class for some of my side projects.

**Comment by bjorn** [comment]
> Mines differs slightly from the others. In Tomáš Andrle's the map must be read in the .CSV format, mines use .XML.

I'm actually deprecating the "XML" tile layer data format in the next version of Tiled due to its shear inefficiency. The "CSV" format is better in terms of both storage space and readability, but of course it requires a bit of extra parsing code.

> E.g you create a reference to a layer within a map, and use that reference to access layer data. My class gives you direct access to any part of a map, by passing the name.

I find the thing about "references" a little confusing since your functions return references as well obviously. I think you basically mean the other libs don't provide a function to search for something by its name?

> Also, in the other loaders, they are designed to load a single map specifically. My class is able to load and them keep any and all maps you load, stored internally in the class so you can access them later at any time. The memory is cleaned up safely later.

Ok, but this feature is just one line like `map<string, unique_ptr<TileMap>>`, which I also see in your code. Though I guess it offers a small bit of convenience.

> My class is read-only.

Why? Sooner or later there will be somebody with a reason to change something. If you don't want to change anything right now, just add `const` when declaring the variable holding the map (but you'll have to mark your getter functions as `const` first).

> However, bjorn if you find there is no benefit in adding mines to the list then that's okay, I will continue using my class for some of my side projects.

Well, I'm a person in favor of avoiding duplication of effort and meaningless choices, but it seems it is impossible in this world. Since this library is more useful to you than the others, the same is probably true for some other people. So I've just added it to the list now. :-)

**Comment by martygrant** [comment]
I'll keep that noted about dropping the XML format, thanks! 

Yeah I could have explained the references part more clearly.

The benefit I saw from a container of maps inside the class was so I could keep all the TMX stuff all wrapped up together in a single package. So I wouldn't need e.g.:

map* level1 = load(...)
map* level2 = load(...)

And then need to pass these maps around to do whatever with them. So I just put them in a container that a user of the class doesn't need to worry about managing. Though I guess it would be just as easy to put the individual maps in your own container if you wanted to.

For being read only, I would want a map to only be edited in Tiled itself. Just to avoid the situation if in one section of code, some part of the map is edited, and then in another section an error occurs because of that data being edited (e.g. accessing data that no longer exists). Though I can see some benefits of being able to change map data during run time too!

I can understand your point. If people prefer the other C++ libraries then fair enough, no need for mine to be there. Though it was a learning experience for me and I picked up some new knowledge, so not a total waste at the end of the day :) Thanks very much for adding it to the list, much appreciated! :)
